### PR TITLE
Remove IDE heuristic for IsCompilationEndAnalyzer

### PR DESCRIPTION
**Pull request by mavasani**
Addresses the core functionality part of https://github.com/dotnet/roslyn/issues/47754#issuecomment-693699421.

A follow-up change will address the second part of generating an IDE-only diagnostic when an analyzer reports a diagnostic in CompilationEnd action which does not have the CompilationEnd custom tag.

### CODE REVIEW

**Comment by sharwell** [comment]
@mavasani Can you fill in this table?

| | IDE Warning without FSA | IDE Warning with FSA |
| --- | --- | --- |
| Analyzer without custom tag | ❓ | ❓ |
| Analyzer with custom tag | ❓ | ❓ |

**Comment by mavasani** [comment]
@sharwell That is not in scope of this PR - this PR just changes the existing logic for determining if a diagnostic reported from build should be treated as a build-only/compilation-end diagnostic or not. Before this PR, that determination was done by an expensive, fragile heuristic that involves invoking callbacks into the analyzer. After this PR, that determination is done via an explicit custom tag on the diagnostic descriptors registered by the analyzer.

**Comment by mavasani** [comment]
To clarify my comment further - FSA setting is not being used to determine if a build diagnostic determined to be a compilation-end diagnostic should be treated as an intellisense supported diagnostic or not (probably it should be). However, those semantics are not being changed in this PR, so changing anything w.r.t. FSA setting is completely orthogonal to any work being done to resolve #47754

**Comment by sharwell** [comment]
@mavasani That's fine (and would be fine to indicate in the table what the follow-up behavior will be). I'm mostly trying to understand the impact this will have on users in each of the four states listed above. For each of the four states, the following could be impacted:

* Behavior after a new solution/file is opened for editing (no prior build or analysis)
* Behavior when Run Code Analysis is invoked in the IDE
* Behavior when a file is opened after a build reported compilation end diagnostics
* Behavior when an analyzer reports diagnostics from normal callbacks and separately registers a compilation end callback

**Comment by mavasani** [comment]
Filed https://github.com/dotnet/roslyn/issues/48793 to track investigating and fixing any build/live de-duping in presence of FSA being on/off. That option is unrelated to this PR and https://github.com/dotnet/roslyn/issues/47754 in general.

**Comment by mavasani** [comment]
@sharwell I performed bunch of validations and hopefully the results below should give us much stronger confidence in the experience:

## Current Experience

1. FSA enabled: Compilation end diagnostics behave just like regular non-compilation end diagnostics. They are converted from build diagnostics to intellisense diagnostics during build/live de-duping and get properly refreshed by background analysis while editing code.
2. FSA disabled: Compilation end diagnostics are build-only diagnostics, that only get refreshed on explicit build.

## New Experience

### Analyzer package with custom tag

Exactly identical to `Current Experience` above.

### Analyzer package without custom tag

1. FSA enabled: Exactly same as `Current Experience` above.
2. FSA disabled: Compilation end diagnostics are reported on build, but only remain in the error list for closed files. As soon as file is opened, these diagnostics disappear from the error list in "Build + Intellisense" setting. The diagnostics still appear in error list when switched to "Build only" setting.

 
2. FSA disabled + Run code analysis command:
| | IDE Warning without FSA | IDE Warning with FSA |
| --- | --- | --- |
| Analyzer without custom tag | ❓ | ❓ |
| Analyzer with custom tag | ❓ | ❓ |

File: src/Compilers/Core/Portable/Diagnostic/WellKnownDiagnosticTags.cs [hunk 1]
```diff
@@ -48,5 +48,11 @@ public static class WellKnownDiagnosticTags
         /// specified by the 'DiagnosticId' property on 'ObsoleteAttribute'.
         /// </summary>
         public const string CustomObsolete = nameof(CustomObsolete);
+
+        /// <summary>
+        /// Indicates that the diagnostic is a compilation end diagnostic reported
+        /// from a compilation end action.
+        /// </summary>
+        public const string CompilationEnd = nameof(CompilationEnd);
```
Comment (mavasani) [inline src/Compilers/Core/Portable/Diagnostic/WellKnownDiagnosticTags.cs:56]: @dotnet/roslyn-compiler This is the only compiler change in this PR (along with the corresponding public API change below). Please see https://github.com/dotnet/roslyn/issues/47754 for more context.

**Review by cston — APPROVED** [review]
Compiler changes LGTM.

**Review by 333fred — APPROVED** [review]
Compiler changes LGTM (commit 7)

File: src/Compilers/Core/Portable/Diagnostic/WellKnownDiagnosticTags.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/Diagnostic/WellKnownDiagnosticTags.cs:56]: Plan is to mark this obsolete (unused) if we ever re-implement the deduplication algorithm to not need it.

File: src/EditorFeatures/CSharpTest/Diagnostics/Suppression/RemoveUnnecessaryPragmaSuppressionsTests.cs [hunk 2]
```diff
@@ -111,7 +111,8 @@ public override void Initialize(AnalysisContext context)
         protected sealed class CompilationEndDiagnosticAnalyzer : DiagnosticAnalyzer
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/Diagnostics/Suppression/RemoveUnnecessaryPragmaSuppressionsTests.cs:111]: ❔ Do we need any tests for compilation end analyzers that _don't_ have the tag?

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_BuildSynchronization.cs [hunk 3]
```diff
@@ -89,7 +89,7 @@ public async Task SynchronizeWithBuildAsync(ImmutableDictionary<ProjectId, Immut
                     }
 
                     // Enqueue re-analysis of open documents.
-                    AnalyzerService.Reanalyze(Workspace, documentIds: Workspace.GetOpenDocumentIds(), highPriority: true);
+                    AnalyzerService.Reanalyze(Workspace, documentIds: Workspace.GetOpenDocumentIds());
```
Comment (sharwell) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_BuildSynchronization.cs:92]: ❔ Why this change?

File: src/EditorFeatures/CSharpTest/Diagnostics/Suppression/RemoveUnnecessaryPragmaSuppressionsTests.cs (hunk 2 above)
Comment (mavasani) [reply src/EditorFeatures/CSharpTest/Diagnostics/Suppression/RemoveUnnecessaryPragmaSuppressionsTests.cs:111]: We do - there was a test failure with just the product changes, which had to be fixed with https://github.com/dotnet/roslyn/pull/48762/commits/e9ee72962adf689093fe9ec87e13bf0be32c6e7c

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_BuildSynchronization.cs (hunk 3 above)
Comment (mavasani) [reply src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_BuildSynchronization.cs:92]: Ah, let me revert - did not intend it here.

Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_BuildSynchronization.cs:92]:
```suggestion
                    AnalyzerService.Reanalyze(Workspace, documentIds: Workspace.GetOpenDocumentIds(), highPriority: true);
```

File: src/EditorFeatures/CSharpTest/Diagnostics/Suppression/RemoveUnnecessaryPragmaSuppressionsTests.cs (hunk 2 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/Diagnostics/Suppression/RemoveUnnecessaryPragmaSuppressionsTests.cs:111]: Right, by adding the tag it looks like we no longer have tests for the case _without_ the tag. I would have expected the fix in e9ee729 to instead be implemented by:

1. Creating a copy of the current `CompilationEndDiagnosticAnalyzer` and adding the tag there
2. Adding a second test alongside the one that failed, where the new test references the new analyzer from (1) and the old test has its assertions updated to account for the fact that the old analyzer doesn't have the new tag

**Review by sharwell — APPROVED** [review]
Looks good aside from the concern of a possible test gap

**Comment by mavasani** [comment]
Thanks @sharwell - I have added the requested test.
